services/queue/src/api.js

```diff
@@ -300,6 +300,100 @@ builder.declare({
   return res.reply(result);
 });
 
+/** Get task group info */
+builder.declare({
+  method: 'get',
+  route: '/task-group/:taskGroupId',
+  name: 'getTaskGroup',
+  scopes: 'queue:list-task-group:<taskGroupId>',
+  stability: APIBuilder.stability.stable,
+  category: 'Tasks',
+  output: 'task-group-response.yml',
+  title: 'Get Task Group',
+  description: [
+    'Get task group information by `taskGroupId`.',
+    '',
+    'This will return meta-information associated with the task group.',
+    'It contains information about task group expiry date or if it is sealed.',
+  ].join('\n'),
+}, async function (req, res) {
+  let taskGroupId = req.params.taskGroupId;
+
+  const taskGroup = await TaskGroup.get(this.db, taskGroupId);
+  if (!taskGroup) {
+    return res.reportError('ResourceNotFound',
+      'No task-group with taskGroupId: `{{taskGroupId}}`', {
+        taskGroupId,
+      });
+  }
+
+  // fetch project ids to construct scopes: `queue:seal-task-group:<taskGroupId>`
+  let projectIds = await taskGroup.getProjectIds(this.db);
+
+  await req.authorize({
+    taskGroupId,
+    projectIds,
+  });
+
+  return res.reply(taskGroup.serialize());
+});
+
+/** Seal Task Group */
+builder.declare({
+  method: 'post',
+  route: '/task-group/:taskGroupId/seal',
+  name: 'sealTaskGroup',
+  scopes: {
+    AnyOf: [
+      'queue:seal-task-group:<taskGroupId>',
+      {
+        AllOf: [{
+          for: 'projectId',
+          in: 'projectIds',
+          each: 'queue:seal-task-group-in-project:<projectId>',
+        }],
+      },
+    ],
+  },
+  stability: APIBuilder.stability.experimental,
+  category: 'Tasks',
+  input: undefined,
+  output: 'task-group-response.yml',
+  title: 'Seal Task Group',
+  description: [
+    'Seal task group to prevent creation of new tasks.',
+    '',
+    'Task group can be sealed once and is irreversible. Calling it multiple times',
+    'will return same result and will not update it again.',
+  ].join('\n'),
+}, async function (req, res) {
+  const taskGroupId = req.params.taskGroupId;
+
+  const taskGroup = await TaskGroup.get(this.db, taskGroupId);
+  if (!taskGroup) {
+    return res.reportError('ResourceNotFound',
+      'No task-group with taskGroupId: `{{taskGroupId}}`', {
+        taskGroupId,
+      });
+  }
+
+  // fetch project ids to construct scopes: `queue:seal-task-group:<taskGroupId>`
+  let projectIds = await taskGroup.getProjectIds(this.db);
+
+  await req.authorize({
+    taskGroupId,
+    projectIds,
+  });
+
+  const updated = TaskGroup.fromDbRows(await this.db.fns.seal_task_group(taskGroupId));
```

Comment:
There is a race condition here. If more projects are added to the task group between fetching the projectIds and updating the database, then there will be a privilege escalation.

Reply:
yes, we'll have to ignore the project part for now. Somehow multiple projects beg for multiple task ids, and we should really force having it so

Reply:
For onlookers - with "ignore the project part", Yarik means we have decided to change the required scopes to just `queue:seal-task-group:<taskGroupId>` i.e. no scopes based on `projectId`. It doesn't mean we are ignoring a privilege escalation path! The privilege escalation path has been removed by requiring stricter scopes. :-)